src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs

```diff
@@ -206,50 +203,64 @@ public void RemoveDelegate(Delegate d, bool wrapArgs = false)
         {
-            lock (_delegateWrappers)
+            DelegateWrapper[] wrappers;
+            DelegateWrapper[] newWrappers;
+            do
             {
-                // Find delegate wrapper indexes. Iterate in reverse such that the list to remove is sorted by high to low index.
-                List<int> toRemove = new List<int>();
-                for (int i = _delegateWrappers.Count - 1; i >= 0; i--)
+                wrappers = _delegateWrappers;
+
+                List<int> toRemove = new();
+                for (int i = 0; i < wrappers.Length; i++)
                 {
-                    DelegateWrapper wrapper = _delegateWrappers[i];
-                    Delegate[] invocationList = wrapper.Delegate.GetInvocationList();
-                    foreach (Delegate delegateMaybe in invocationList)
+                    DelegateWrapper wrapper = wrappers[i];
+                    if (condition(wrapper.Delegate))
                     {
-                        if (condition(delegateMaybe))
-                        {
-                            Delegate? newDelegate = Delegate.Remove(wrapper!.Delegate, delegateMaybe);
-                            if (newDelegate != null)
-                            {
-                                wrapper.Delegate = newDelegate;
-                            }
-                            else
-                            {
-                                toRemove.Add(i);
-                            }
-                        }
+                        toRemove.Add(i);
                     }
                 }
 
-                foreach (int idx in toRemove)
+                if (toRemove.Count == 0)
                 {
-                    _delegateWrappers.RemoveAt(idx);
+                    // Nothing to remove
+                    return;
                 }
+
+                newWrappers = RemoveAll(wrappers, CollectionsMarshal.AsSpan(toRemove));
+            } while (!PublishNewWrappers(newWrappers, wrappers));
+
+            // The list of indices is assumed to be sorted in ascending order
+            static DelegateWrapper[] RemoveAll(DelegateWrapper[] oldCol, ReadOnlySpan<int> toRemove)
+            {
+                // Allocate new collection
+                var newCol = new DelegateWrapper[oldCol.Length - toRemove.Length];
+                if (newCol.Length == 0)
+                {
+                    return newCol;
+                }
+
+                // Iterate over collection, skipping elements that should be removed.
+                int ri = 0;
+                for (int oi = 0, ni = 0; oi < oldCol.Length; oi++)
+                {
+                    if (ri < toRemove.Length && oi == toRemove[ri])
+                    {
+                        ri++;
+                        continue;
+                    }
+
+                    newCol[ni] = oldCol[oi];
+                    ni++;
+                }
+
+                return newCol;
```

Comment:
Seems like this whole thing could just be:
```C#
DelegateWrapper[] oldWrappers, newWrappers;
do
{
    oldWrappers = _delegateWrappers;
    var tmp = new List<DelegateWrapper>(oldWrappers);
    tmp.RemoveAll(w => condition(w.Delegate));
    newWrappers = tmp.ToArray();
}
while (!PublishNewWrappers(newWrappers, oldWrappers));
```
I'm just not clear on what overheads we're trying to avoid and which we're not.